packages/react-components/react-color-picker-preview/library/src/components/ColorArea/useColorArea.ts

```diff
@@ -9,23 +18,143 @@ import type { ColorAreaProps, ColorAreaState } from './ColorArea.types';
  * before being passed to renderColorArea_unstable.
  *
  * @param props - props from this instance of ColorArea
- * @param ref - reference to root HTMLDivElement of ColorArea
+ * @param ref - reference to root HTMLInputElement of ColorArea
  */
 export const useColorArea_unstable = (props: ColorAreaProps, ref: React.Ref<HTMLDivElement>): ColorAreaState => {
-  return {
-    // TODO add appropriate props/defaults
+  const { targetDocument } = useFluent();
+  const rootRef = React.useRef<HTMLDivElement>(null);
+  const xRef = React.useRef<HTMLInputElement>(null);
+  const yRef = React.useRef<HTMLInputElement>(null);
+  const focusWithinRef = useFocusWithin();
+
+  const {
+    onChange,
+    // Slots
+    inputX,
+    inputY,
+    thumb,
+
+    ...rest
+  } = props;
+
+  const [color, setColor] = useControllableState<HsvColor>({
+    defaultState: props.defaultColor,
+    state: props.color,
+    initialState: INITIAL_COLOR_HSV,
+  });
+
+  const saturation = Math.round(color.s * 100);
+  const value = Math.round(color.v * 100);
+
+  const requestColorChange = useEventCallback((event: MouseEvent) => {
+    if (!rootRef.current) {
+      return;
+    }
+    const coordinates = getCoordinates(rootRef.current, event);
+    const newColor: HsvColor = {
+      ...color,
+      s: coordinates.x,
+      v: coordinates.y,
+    };
+
+    setColor(newColor);
+    onChange?.(event, { type: 'change', event, color: newColor });
+  });
+
+  const handleDocumentMouseMove = React.useCallback(
+    (event: MouseEvent) => {
+      requestColorChange(event);
+    },
+    [requestColorChange],
+  );
+  const handleDocumentMouseUp = useEventCallback(() => {
+    targetDocument?.removeEventListener('mousemove', handleDocumentMouseMove);
+  });
+
+  const _onMouseDown: React.MouseEventHandler<HTMLDivElement> = useEventCallback(event => {
+    event.stopPropagation();
+    event.preventDefault();
+
+    requestColorChange(event.nativeEvent);
+
+    targetDocument?.addEventListener('mousemove', handleDocumentMouseMove);
+    targetDocument?.addEventListener('mouseup', handleDocumentMouseUp, { once: true });
+  });
+
+  const _onChange: React.ChangeEventHandler<HTMLInputElement> = useEventCallback(event => {
+    const newColor = { h: color.h, s: Number(xRef.current!.value) / 100, v: Number(yRef.current?.value) / 100, a: 1 };
+
+    setColor(newColor);
+    onChange?.(event, {
+      type: 'change',
+      event,
+      color: newColor,
+    });
+  });
```

Comment:
I suggest to have to separate handlers for `x` & `y`, so you won't deal with refs at all and would be able to rely on `event.target.event`